File: mt5-core-server/src/services/auto/analyzers/smc/orderBlocks.ts
/**
 * SMC Engine V2 — Order Block Detection
 * Port จาก Pine Script V8.3: Strict SMC + FVG Confirm + Mitigation
 * 2026-05-06
 */

import type { Candle, FVG, OrderBlock, StructureDirection } from './types.js';
import { detectSwings, candleMax, candleMin, isBearish, isBullish } from './utils.js';

export interface OrderBlockConfig {
  swingLookback: number;
  useBody: boolean;
  filterByTrend: boolean;
  maxOBCount: number;
}

const DEFAULT_CFG: OrderBlockConfig = {
  swingLookback: 5,
  useBody: true,
  filterByTrend: true,
  maxOBCount: 30,
};

export function detectOrderBlocks(
  candles: Candle[],
  fvgs: FVG[],
  structDir: StructureDirection = 'INIT',
  config: Partial<OrderBlockConfig> = {}
): { bullOBs: OrderBlock[]; bearOBs: OrderBlock[] } {
  const cfg = { ...DEFAULT_CFG, ...config };
  if (candles.length < 20) return { bullOBs: [], bearOBs: [] };

  const { highs, lows } = detectSwings(candles, cfg.swingLookback);
  const bullOBs: OrderBlock[] = [];
  const bearOBs: OrderBlock[] = [];
  const gMax = (c: Candle) => cfg.useBody ? candleMax(c) : c.h;
  const gMin = (c: Candle) => cfg.useBody ? candleMin(c) : c.l;

  // Bullish OBs
  for (const swing of highs) {
    const breakIdx = candles.findIndex((c, idx) => idx > swing.index && c.c > swing.price);
    if (breakIdx === -1) continue;
    let lowestLow = gMin(candles[swing.index]), lowestIdx = swing.index;
    for (let i = swing.index; i < breakIdx; i++) {
      const low = gMin(candles[i]);
      if (low <= lowestLow) { lowestLow = low; lowestIdx = i; }
    }
    let obIdx = lowestIdx, found = false;
    for (let k = lowestIdx; k <= Math.min(breakIdx + 5, candles.length - 1); k++) {
      if (isBearish(candles[k])) { obIdx = k; found = true; break; }
    }
    if (!found) continue;
    if (cfg.filterByTrend && structDir === 'BEARISH') continue;
    const top = gMax(candles[obIdx]), btm = gMin(candles[obIdx]);
    const hasFVG = fvgs.some(f => f.type === 'BULL' && Math.abs(f.barIndex - obIdx) <= 5);
    if (hasFVG) {
      bullOBs.unshift({ type: 'BULL', top, bottom: btm, barIndex: obIdx, mitigated: false, hasFVG, invalidated: false, volume: candles[obIdx].v });
    }
  }

  // Bearish OBs
  for (const swing of lows) {
    const breakIdx = candles.findIndex((c, idx) => idx > swing.index && c.c < swing.price);
    if (breakIdx === -1) continue;
    let highestHigh = gMax(candles[swing.index]), highestIdx = swing.index;
    for (let i = swing.index; i < breakIdx; i++) {
      const high = gMax(candles[i]);
      if (high >= highestHigh) { highestHigh = high; highestIdx = i; }
    }
    let obIdx = highestIdx, found = false;
    for (let k = highestIdx; k <= Math.min(breakIdx + 5, candles.length - 1); k++) {
      if (isBullish(candles[k])) { obIdx = k; found = true; break; }
    }
    if (!found) continue;
    if (cfg.filterByTrend && structDir === 'BULLISH') continue;
    const top = gMax(candles[obIdx]), btm = gMin(candles[obIdx]);
    const hasFVG = fvgs.some(f => f.type === 'BEAR' && Math.abs(f.barIndex - obIdx) <= 5);
    if (hasFVG) {
      bearOBs.unshift({ type: 'BEAR', top, bottom: btm, barIndex: obIdx, mitigated: false, hasFVG, invalidated: false, volume: candles[obIdx].v });
    }
  }

  // Update mitigation & invalidation
  for (const ob of bullOBs) {
    for (let j = ob.barIndex + 1; j < candles.length; j++) {
      if (candles[j].l <= ob.top && candles[j].h >= ob.bottom) { ob.mitigated = true; break; }
    }
    if (candleMin(candles[candles.length - 1]) < ob.bottom) ob.invalidated = true;
  }
  for (const ob of bearOBs) {
    for (let j = ob.barIndex + 1; j < candles.length; j++) {
      if (candles[j].h >= ob.bottom && candles[j].l <= ob.top) { ob.mitigated = true; break; }
    }
    if (candleMax(candles[candles.length - 1]) > ob.top) ob.invalidated = true;
  }

  return {
    bullOBs: bullOBs.filter(ob => !ob.invalidated).slice(0, cfg.maxOBCount),
    bearOBs: bearOBs.filter(ob => !ob.invalidated).slice(0, cfg.maxOBCount),
  };
}

File: composeApp/src/commonMain/kotlin/com/example/personalaibot/automation/smc/OrderBlocks.kt
package com.example.personalaibot.automation.smc

import com.example.personalaibot.tools.trading.Candle
import kotlin.math.abs
import kotlin.math.min

/**
 * Order Block Detection — port จาก mt5-core-server orderBlocks.ts (Pine V8.3)
 * Strict SMC: ต้องมี FVG confirm (displacement) + กรองตามทิศโครงสร้าง
 * คืน list เรียง "ใหม่สุดก่อน" (index 0 = ล่าสุด) ตรงกับต้นฉบับที่ใช้ unshift
 */
internal object OrderBlockDetection {

    fun detect(
        candles: List<Candle>,
        fvgs: List<Fvg>,
        structDir: StructureDirection = StructureDirection.INIT,
        swingLookback: Int = 5,
        useBody: Boolean = true,
        filterByTrend: Boolean = true,
        maxOBCount: Int = 30
    ): Pair<List<SmcOrderBlock>, List<SmcOrderBlock>> {
        if (candles.size < 20) return emptyList<SmcOrderBlock>() to emptyList()

        val (highs, lows) = detectSwings(candles, swingLookback)
        val bullOBs = mutableListOf<SmcOrderBlock>()
        val bearOBs = mutableListOf<SmcOrderBlock>()
        val gMax = { c: Candle -> if (useBody) candleMax(c) else c.high }
        val gMin = { c: Candle -> if (useBody) candleMin(c) else c.low }

        // Bullish OBs: swing high ที่ถูกทะลุขึ้น → OB = แท่ง bearish แรกหลังจุดต่ำสุดในช่วง
        for (swing in highs) {
            var breakIdx = -1
            for (idx in swing.index + 1 until candles.size) {
                if (candles[idx].close > swing.price) { breakIdx = idx; break }
            }
            if (breakIdx == -1) continue

            var lowestLow = gMin(candles[swing.index])
            var lowestIdx = swing.index
            for (i in swing.index until breakIdx) {
                val low = gMin(candles[i])
                if (low <= lowestLow) { lowestLow = low; lowestIdx = i }
            }
            var obIdx = -1
            for (k in lowestIdx..min(breakIdx + 5, candles.size - 1)) {
                if (isBearish(candles[k])) { obIdx = k; break }
            }
            if (obIdx == -1) continue
            if (filterByTrend && structDir == StructureDirection.BEARISH) continue

            val hasFvg = fvgs.any { it.isBull && abs(it.barIndex - obIdx) <= 5 }
            if (hasFvg) {
                bullOBs.add(0, SmcOrderBlock(
                    isBull = true, top = gMax(candles[obIdx]), bottom = gMin(candles[obIdx]),
                    barIndex = obIdx, hasFvg = true, volume = candles[obIdx].volume
                ))
            }
        }

        // Bearish OBs: swing low ที่ถูกทะลุลง → OB = แท่ง bullish แรกหลังจุดสูงสุดในช่วง
        for (swing in lows) {
            var breakIdx = -1
            for (idx in swing.index + 1 until candles.size) {
                if (candles[idx].close < swing.price) { breakIdx = idx; break }
            }
            if (breakIdx == -1) continue

            var highestHigh = gMax(candles[swing.index])
            var highestIdx = swing.index
            for (i in swing.index until breakIdx) {
                val high = gMax(candles[i])
                if (high >= highestHigh) { highestHigh = high; highestIdx = i }
            }
            var obIdx = -1
            for (k in highestIdx..min(breakIdx + 5, candles.size - 1)) {
                if (isBullish(candles[k])) { obIdx = k; break }
            }
            if (obIdx == -1) continue
            if (filterByTrend && structDir == StructureDirection.BULLISH) continue

            val hasFvg = fvgs.any { !it.isBull && abs(it.barIndex - obIdx) <= 5 }
            if (hasFvg) {
                bearOBs.add(0, SmcOrderBlock(
                    isBull = false, top = gMax(candles[obIdx]), bottom = gMin(candles[obIdx]),
                    barIndex = obIdx, hasFvg = true, volume = candles[obIdx].volume
                ))
            }
        }

        // mitigation & invalidation
        val last = candles[candles.size - 1]
        for (ob in bullOBs) {
            for (j in ob.barIndex + 1 until candles.size) {
                if (candles[j].low <= ob.top && candles[j].high >= ob.bottom) { ob.mitigated = true; break }
            }
            if (candleMin(last) < ob.bottom) ob.invalidated = true
        }
        for (ob in bearOBs) {
            for (j in ob.barIndex + 1 until candles.size) {
                if (candles[j].high >= ob.bottom && candles[j].low <= ob.top) { ob.mitigated = true; break }
            }
            if (candleMax(last) > ob.top) ob.invalidated = true
        }

        return bullOBs.filter { !it.invalidated }.take(maxOBCount) to
                bearOBs.filter { !it.invalidated }.take(maxOBCount)
    }
}
